lms/static/scripts/frontend_apps/components/LMSPagePicker.tsx

```diff
@@ -0,0 +1,300 @@
+      There are no pages in this instance.{' '}
+      <a href={'TODO'} target="_blank" rel="noreferrer">
+        Create some new pages
+      </a>{' '}
+      and try again.
+    </div>
+  );
+}
+
+type LMSPagePickerProps = {
+  /**
+   * Auth token for use in calls to the backend
+   */
+  authToken: string;
+
+  /**
+   * Config for the API call to list available pages
+   */
+  listPagesApi: APICallInfo;
+
+  /**
+   * Callback invoked if the user cancels page selection
+   */
+  onCancel: () => any;
+
+  /**
+   * Callback invoked with the metadata of the selected page if the user makes
+   * a selection
+   */
+  onSelectPage: (p: Page) => void;
+};
+
+type FetchingState = {
+  state: 'fetching';
+};
+
+type FetchedState = {
+  state: 'fetched';
+  pages: Page[];
+};
+
+type AuthorizingState = {
+  state: 'authorizing';
+
+  /**
+   * Flag indicating that authorization has previously been attempted
+   */
+  isRetry: boolean;
+};
+
+type ErrorState = {
+  state: 'error';
+  error: Error;
+};
+
+type LMSPagePickerState =
+  | FetchingState
+  | FetchedState
+  | AuthorizingState
+  | ErrorState;
+
+type AuthorizeAction = {
+  type: 'authorize';
+  label: string;
+};
+
+type ReloadAction = {
+  type: 'reload';
+};
+
+type SelectAction = {
+  type: 'select';
+  label: string;
+  disabled: boolean;
+};
+
+type ContinueAction = AuthorizeAction | ReloadAction | SelectAction;
+
+/**
+ * A page picker dialog that allows the user to choose pages from their
+ * LMS.
+ *
+ * The picker will attempt to list pages when mounted, and will show an
+ * authorization popup if necessary.
+ */
+export default function LMSPagePicker({
+  authToken,
+  listPagesApi,
+  onCancel,
+  onSelectPage,
+}: LMSPagePickerProps) {
+  const [dialogState, setDialogState] = useState<LMSPagePickerState>({
+    state: 'fetching',
+  });
+
+  // Has the first attempt to fetch the list of pages in the LMS completed?
+  const [initialFetch, setInitialFetch] = useState(true);
+  const [selectedPage, setSelectedPage] = useState<Page | null>(null);
+
+  const loadPagesToDisplay = useCallback(
+    async (): Promise<Page[]> =>
+      apiCall({
+        authToken,
+        path: listPagesApi.path,
+      }),
+    [authToken, listPagesApi]
+  );
+  const computePagesToDisplay = useCallback(
+    /**
+     * Load the list of pages to display in the picker. Handle error states if they arise.
+     *
+     * @param isReload - When true, request a fresh fetch of pages from the API
+     */
+    async (isReload = false) => {
+      try {
+        setDialogState({ state: 'fetching' });
+        const pages = await loadPagesToDisplay();
+        setDialogState({ state: 'fetched', pages });
+      } catch (error) {
+        if (isAuthorizationError(error)) {
+          setDialogState({ state: 'authorizing', isRetry: isReload });
+        } else {
+          setDialogState({ state: 'error', error });
+        }
+      }
+      setInitialFetch(false);
+    },
+    [loadPagesToDisplay]
```

Comment:
These two function could probably be merged. In the file picker it makes more sense to have them separated as the logic is more complex.